src/urql/atomWithQuery.ts

```diff
@@ -0,0 +1,132 @@
+  >(
+    new Promise<OperationResult<Data, Variables>>(() => {}) // infinite pending
+  )
+  const queryAtom = atom<
+    [QueryArgs<Data, Variables>, WritableAtom<null, any>],
+    ResultActions
+  >(
+    (get) => {
+      const args = createQueryArgs(get)
+      const initAtom = atom(
+        null,
+        (
+          get,
+          set,
+          action:
+            | { type: 'init'; initializer: (client: Client) => void }
+            | { type: 'result'; result: OperationResult<Data, Variables> }
+        ) => {
+          if (action.type === 'init') {
+            set(
+              operationResultAtom,
+              new Promise<OperationResult<Data, Variables>>(() => {}) // new fetch
+            )
+            if (!args.pause) {
+              action.initializer(getClient(get, set))
+            }
+          } else if (action.type === 'result') {
+            if (!action.result.stale) {
+              set(operationResultAtom, action.result)
+            }
+          }
+        }
+      )
+      initAtom.onMount = (dispatch) => {
+        let unsub: (() => void) | undefined | false
+        const initializer = (client: Client) => {
+          const subscription = pipe(
+            client.query(args.query, args.variables, {
+              requestPolicy: args.requestPolicy,
+              ...args.context,
+            }),
+            subscribe((result) => {
+              dispatch({ type: 'result', result })
+            })
+          )
+          if (unsub === false) {
+            subscription.unsubscribe()
+          } else {
+            unsub = () => {
+              subscription.unsubscribe()
+            }
+          }
+        }
+        dispatch({ type: 'init', initializer })
+        return () => {
+          if (unsub) {
+            unsub()
+          }
+          unsub = false
+        }
+      }
+      return [args, initAtom]
+    },
+    (get, set, action) => {
+      if (action.type === 'execute') {
+        const [args] = get(queryAtom)
+        const client = getClient(get, set)
+        client
+          .query(args.query, args.variables, {
+            requestPolicy: args.requestPolicy,
+            ...args.context,
+            ...action.opts,
+          })
+          .toPromise()
+          .then((result) => {
+            if (!result.stale /* XXX can this happen? */) {
+              set(operationResultAtom, result)
+            }
+          })
+      }
+    }
+  )
+  const queryDataAtom = atom<[Data, GraphQLExtensions], ResultActions>(
+    (get) => {
+      const [, initAtom] = get(queryAtom)
+      get(initAtom) // use it here
+      const operationResult = get(operationResultAtom)
+      if (operationResult.error) {
+        throw operationResult.error
+      }
+      if (operationResult.data === undefined) {
+        // XXX can this happen?
+        throw new Error('no data in operation result')
+      }
+      return [operationResult.data, operationResult.extensions]
```

Comment:
I changed to return a tuple with some regret.
What we lose here is `operationResult.operation`. Is it acceptable?

Reply:
I saw that you swapped this out; just on a side-note, I'd say that _every_ property on `OperationResult` is potentially interesting and important to users, so I'd hand the entire result through as-is.

Reply:
Understand what you mean, but for example `stale` would never be true. Isn't it confusing to have a always-false property? Maybe not for you. We are looking at this from different perspectives.
More issue for me is, `Data` shouldn't be undefined. That's very important for TypeScript users.

Reply:
@dai-shi `stale` would only ever be `false` if you decide to filter it (and I think you did?) but I'm not sure whether it's a good idea to hide this just by virtue of being in "jotai land", since that also means that users don't get access to certain pieces of data anymore, like for instance partial data with schema awareness.

But even then you'd probably want to at least match what's already in the type itself: https://github.com/FormidableLabs/urql/blob/12d49d3feccc830c95ef07f488f27655d4010b58/packages/core/src/types.ts#L63-L74

So you could skip `stale` but all the other properties should be made available.

Reply:
What about error? We already throw it, so it will never be defined.

So, it's like `Required<Pick<OperationResult, 'data'>> & Omit<OperationResult, 'data' | 'error' | 'stale'>`?

Well, I can just return OperationResult, but my strong desire is to make `Data` not optional.

Reply:
Throwing it is unfortunately not an option because we may have partial data with some errors present. I can't find a long explanation for this, but here's a short one: https://github.com/FormidableLabs/urql/discussions/1456#discussioncomment-491031

Reply:
Oh, that's true. Forgot it's graphql 😓 . I'm convinced.

(Curious if we can distinguish graphql error from others like 404, with which we can throw.)